File: library/sync-tasks/src/main/java/org/cru/godtools/sync/task/BaseSyncTasks.java
package org.cru.godtools.sync.task;

import android.content.Context;
import android.os.Bundle;

import org.cru.godtools.api.GodToolsApi;
import org.cru.godtools.model.Base;
import org.greenrobot.eventbus.EventBus;
import org.keynote.godtools.android.db.GodToolsDao;

import java.util.Collection;

import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;
import androidx.collection.LongSparseArray;
import androidx.collection.SimpleArrayMap;

import static android.content.ContentResolver.SYNC_EXTRAS_MANUAL;

@WorkerThread
abstract class BaseSyncTasks {
    final GodToolsApi mApi;
    final GodToolsDao mDao;
    private final EventBus mEventBus;

    BaseSyncTasks(@NonNull final Context context) {
        mApi = GodToolsApi.getInstance();
        mDao = GodToolsDao.Companion.getInstance(context);
        mEventBus = EventBus.getDefault();
    }

    static boolean isForced(@NonNull final Bundle extras) {
        return extras.getBoolean(SYNC_EXTRAS_MANUAL, false);
    }

    @NonNull
    static <E extends Base> LongSparseArray<E> index(@NonNull final Collection<E> items) {
        final LongSparseArray<E> index = new LongSparseArray<>();
        for (final E item : items) {
            index.put(item.getId(), item);
        }
        return index;
    }

    static void coalesceEvent(@NonNull final SimpleArrayMap<Class<?>, Object> events, @NonNull final Object event) {
        final Class<?> type = event.getClass();
        Object currEvent = events.get(type);
        if (currEvent != null) {
            // coalesce any events that need to be coalesced
        } else {
            currEvent = event;
        }
        events.put(type, currEvent);
    }

    void sendEvents(@NonNull final SimpleArrayMap<Class<?>, Object> events) {
        for (int i = 0; i < events.size(); i++) {
            mEventBus.post(events.valueAt(i));
        }
        events.clear();
    }
}

File: library/sync-tasks/src/main/java/org/cru/godtools/sync/task/BaseSyncTasks.kt
package org.cru.godtools.sync.task

import android.content.ContentResolver
import android.os.Bundle
import androidx.annotation.RestrictTo
import androidx.annotation.WorkerThread
import androidx.collection.LongSparseArray
import androidx.collection.SimpleArrayMap
import org.cru.godtools.model.Base
import org.greenrobot.eventbus.EventBus

@WorkerThread
@RestrictTo(RestrictTo.Scope.LIBRARY)
abstract class BaseSyncTasks internal constructor(private val eventBus: EventBus) {
    fun sendEvents(events: SimpleArrayMap<Class<*>, Any>) {
        for (i in 0 until events.size()) eventBus.post(events.valueAt(i))
        events.clear()
    }

    companion object {
        fun isForced(extras: Bundle) = extras.getBoolean(ContentResolver.SYNC_EXTRAS_MANUAL, false)

        fun <E : Base> index(items: Collection<E>) = LongSparseArray<E>().apply {
            for (item in items) put(item.id, item)
        }

        fun coalesceEvent(events: SimpleArrayMap<Class<*>, Any>, event: Any) = events.put(event.javaClass, event)
    }
}
